torch/csrc/jit/codegen/cuda/scheduler/transpose.h

```diff
@@ -0,0 +1,97 @@
+// to as much input and output tensors as possible. For a DAG with only pure
+// pointwise operators, the scheduling is very simple because the inner most
+// dimension of all input and output tensors are all mapped together in the
+// ComputeAtMap, i.e., there is essentially only one inner most dimension. In
+// such case, we just vectorize that inner most dimension and bind it to
+// threadIdx.x identically for all input and output tensors. In the case where
+// transposes are present in the DAG, the inner most dimensions of different
+// inputs and outputs might not match. And there is no fixed pattern on which
+// input/output tensors should share the same inner most dimension with which.
+// Consider the following example DAGs ([T] represents transpose, all tensors
+// are 2D):
+//
+//   t0    t1      t0    t1      t0    t1        t0    t1         t0
+//    \    |        \    /        \    |          \    |          |
+//     \  [T]       [T] [T]        \  [T]          t2 [T]        [T]
+//      \ /           \ /           \ / \         / \ / \         |
+//      t2             t2           t2   t3      t3  t4 t5       [T]
+//                                                                |
+//                                                                t1
+//
+// In order to support all these cases in a general way, the following
+// perspective is very important: What we are looking for is to bind threadIdx.x
+// differently for different inputs and outputs, so there has to be some tensor
+// somewhere in the DAG that we write and read with different threadIdx.x
+// bindings. The tensor of binding swap can be any tensor on the path that
+// connects inputs/outputs with different inner most dimension, especially, it
+// does not necessarily be the tensor of the transpose operator. In other words,
+// thanks to our indexing system who is already taking care of the correctness
+// of transpose, the scheduler can freely choose where to realize these
+// transposes as different threadIdx.x bindings. This observation greatly
+// simplifies our scheduling.
+//
+// Our scheduling strategyis is as follows: We first split the inputs and
+// outputs of the fusion into two groups according to their inner most
+// dimension. The inner most dimensions of tensors in the same group are mapped
+// to each other, and they are not mapped to the inner most dimesion of tensors
+// in a different group. Depending on the transpose pattern, there can be more
+// than two groups, if this is the case, we only consider the two largest
+// groups, and the tensors in the remaining groups will just be accessed
+// unvectorized and uncoalesced. We call the largest group as `group1` and the
+// second largest group as `group2`. When we have the split, we will make a 2D
+// tiling [I1, I2] -> [I1/tile1, tile1, I2/tile2, tile2] on the inner most
+// dimensions of group1 and group2. Each tile [tile1, tile2] will be handled by
```

Comment:
When finding the two largest groups, are we assuming they are transposed?

Reply:
For most of the cases, yes, they should be transposed. But fundamentally, they don't have to be transposed, and I don't care what is the reason why they don't have matching inner most dim. For example, for the following fusion:
```C++
  auto tv0 = makeConcreteTensor({1024, 1, 256});
  auto tv1 = makeConcreteTensor({1024, 1024, 1});
  fusion.addInput(tv0);
  fusion.addInput(tv1);
  auto tv2 = add(tv0, tv1);
  fusion.addOutput(tv2);
```
The transpose scheduler is outperforming the pointwise scheduler. I get 840GB/s (transpose scheduler) vs 800GB/s (pointwise scheduler) on RTX 3090. Note that this problem has no transpose at all. But there are mismatched innermost dimensions due to broadcasting.